### DISCUSSION TITLE
Static site generated bundle - performance for production

### DISCUSSION THREAD

**Discussion by saniko** [vercel/next.js#17361] [Help] [answered]
Hello,

I'm building several static Next.js applications (some of them with ~200 pages).
Can someone please elaborate on the bundling process?, especially about performance, lazy loading (if any), etc.
My aim is to produce a minimal bundle size for each page so I'm planning to include only relevant components.

My question is this, after implementing getStaticProps and running build & export for my project:
Does next.js bundles all components into one file for one HTTP request or do it "lazy loads" relevant chunk per page?
Is there any way to "tree-shake" or remove unused next.js features from the final bundle?
Is there any "best practice" list for optimizing next.js static sites for production?

Thanks for answering.

**Comment by lfades** [comment, marked as answer]
Next.js takes care of all of what the optimizations you're looking for. Take a look at our documentation for more: https://nextjs.org/docs/basic-features/pages

The build output can help you know the size of your pages, devtools and lighthouse are also very good tools if you want to the final performance of your application. But the TL;DR is: Next.js will do everything you just said for you.

**Reply by saniko** [reply]
Thank you for your answer.
I will try to refine my question, 

let's say I have 2 pages.
Page1 uses 2 components, A and B.
Page2 uses 3 components, A, C, D

Will I get a final bundle containing all 3 components or will I get 2 bundles per page,
that will lazy load the relevant components per page?

I'm asking that because I'm planning to use a code generator to create pages dynamically from CMS data,
Should I optimize per page or next will "scan" my entire site and bundle all components for one HTTP request, regardless of where I use them?.

Thanks.

**Reply by lfades** [reply]
Next.js has per page bundles and will create chunks depending on the libraries and components that are being used in pages.

**Reply by lfades** [reply]
Components and libraries can be lazy loaded with [dynamic import and next/dynamic](https://nextjs.org/docs/advanced-features/dynamic-import)

**Reply by saniko** [reply]
Thanks,
But we have checked and when using dynamic import and next/dynamic, we do not get the full HTML in the final bundle.
Apparently, it gets created in runtime?

**Reply by saniko** [reply]
That is why we plan
 to use a code generator

**Reply by saniko** [reply]
If I use Dynamic Import, should I get the static HTML after the build?
Because currently, from what we have checked we do not get the full HTML in the final static output.
It is created at run time.

I must provide to my client 100% static site fro SEO purposes

**Reply by lfades** [reply]
If that's the case then don't use Dynamic Import. Next.js takes care of only building the components and libraries that are required for the page, on a per-page basis. You won't be shipping unrequired JS.

Dynamic Import can help you lazy loading components but based on what you're saying that's not what you're looking for.
